Nodes and clients Merge update (#7133)
ethereum/ethereum-org-website: src/content/developers/docs/nodes-and-clients/index.md

## What are nodes and clients? {#what-are-nodes-and-clients}

"Node" refers to a running piece of client software. A client is an implementation of Ethereum that verifies all transactions in each block, keeping the network secure and the data accurate.

> An Ethereum node consists of two pieces of software: execution client and consensus client. 

Author: I followed "consensus client, execution client" terminology used in current docs but I agree it can be confusing. I will add info on various terminology to give better context

## What are nodes and clients? {#what-are-nodes-and-clients}

"Node" refers to a running piece of client software. A client is an implementation of Ethereum that verifies all transactions in each block, keeping the network secure and the data accurate.

> Post-Merge Ethereum consists of two networks: the execution layer and the consensus layer. Both networks are run by different client software. On this page, we'll refer to them as the execution client and consensus client. 

Reviewer: > Post-Merge Ethereum consists of two networks

_Should_ we distinguish it as "two networks"? Makes things a bit confusing IMO to say that Ethereum will now be "two networks". Could we say it's two layers of a single network? Is there actually a technical difference that makes claiming "two networks" more accurate/clear?

Reviewer: From the EL/CL bullet points below, it seems like we refer to "the network", which further confuses me. Seems like we could stick with Ethereum being a single network but LMK what you think @taxmeifyoucan.

Reviewer 2: Agree, neither can function on its own after The Merge

Author: Agreed, it's more confusing than I realized. I used the word `networks` because they both have own p2p network and I wanted to emphasize that Beacon Chain  is not a new thing and the network has been running for a long time. 

I would rephrase it as layers - `Post-Merge Ethereum consists of two parts: the execution layer and the consensus layer. Both layers are run by different client software.`

## What are nodes and clients? {#what-are-nodes-and-clients}

> "Node" refers to a running piece of client software. A client is an implementation of Ethereum that verifies all transactions in each block, keeping the network secure and the data accurate.

Reviewer 3: > "Node" refers to a running piece of client software.

This sentence doesn't make sense. IMO here we should describe **what a node is** and **not what running a node is** (as per the heading)

Author: Makes sense. Changing to `"Node" refers to a running piece of client software.`

Author: Fair enough, I am wondering how to define it properly. What would you suggest?  
"Running" here is meant as client software which is running/executed somewhere, connected to the network, etc, not the act of somebody running it. Maybe something like `A "node" is an instance of Ethereum client participating in the network.`

Reviewer 2: Yeah, fundamentally a node is Ethereum software being run on a computer, and that software communicates with others doing the same. 

Perhaps this?
`A "node" is any instance of Ethereum client software that is connected to other computers also running Ethereum software, forming a network.`

> A client is an implementation of Ethereum that verifies all transactions in each block, keeping the network secure and the data accurate.

Note here not all clients deal with verifying transactions, ie consensus client, as you get into below... Maybe we make this a little more agnostic?
`A client is an implementation of Ethereum that verifies data against the protocol rules and keeps the network secure.`

Author: @wackerow Sounds good to me!

Before [The Merge](/upgrades/merge/), consensus and execution layer networks were separeted. All transactions and user activity on Ethereum network was happening at what is now the execution layer. One client software provided both execution enviroment and consensus verification of blocks produced by miners. 
The consensus layer, [the Beacon Chain](/upgrades/beacon-chain/), was running separetly since December 2020. It introduced proof-of-stake and coordinated the network of validators based on data from the Ethereum network. 

With the Merge, Ethereum network transitions to proof-of-stake by connecting these networks. Execution and consensus client work together to verify blockchain state. 

Modular design with various pieces of software working together can be referred to as [encapusalted complexity](https://vitalik.ca/general/2022/02/28/complexity.html). This approach makes it easier to execute The Merge in a seamless way and also enables reuse of individual clients, for example in the L2 ecosystem. 

![Coupled execution and consensus clients](./eth1eth2client.png)
Simplified diagram of a coupled execution and consensus client.

### Client diversity {#client-diversity}

> Both [Execution clients](/developers/docs/nodes-and-clients/#consensus-clients) and [Conesnsus clients](/developers/docs/nodes-and-clients/#execution-clients) exist, in a variety of programming languages such as Go, Rust, JavaScript, Typescript, Python, C# .NET, Nim and Java. 

Reviewer 2: ```suggestion
Both [execution clients](/developers/docs/nodes-and-clients/#consensus-clients) and [consensus clients](/developers/docs/nodes-and-clients/#execution-clients) exist, in a variety of programming languages such as Go, Rust, JavaScript, Typescript, Python, C# .NET, Nim and Java. 
```

> Multiple client implementations make the network stronger and more diverse. The ideal goal is to achieve diversity without any client dominating, thereby reducing any single points of failure. The variety of languages also invites a broader developer community and allows them to create integrations in their preferred language.

Author: How about adding `...more diverse than relying on a single client.`?

> Multiple client implementations make the network stronger and more diverse. The ideal goal is to achieve diversity without any client dominating, thereby reducing any single points of failure. The variety of languages also invites a broader developer community and allows them to create integrations in their preferred language.

Reviewer 3: Sorry, I was unclear 😅. 

Describing having multiple client implementations as being 'diverse' has no meaning unless you understand what client diversity means. But, at this point in the text we haven't explained client diversity yet, so using the word 'diverse' might not make sense.

> Multiple client implementations make the network stronger and more diverse. The ideal goal is to achieve diversity without any client dominating, thereby reducing any single points of failure. The variety of languages also invites a broader developer community and allows them to create integrations in their preferred language.

Author: Thanks for clarifying! Tried to rephrase it, please suggest a change if it still doesn't sound right: 
`Multiple client implementations can make the network stronger by reducing its dependency on a single codebase. The ideal goal is to achieve diversity without any client dominating the network, thereby eliminating a potential single point of failure.`

> Learn more about [Client diversity](/developers/docs/client-diversity). 

Reviewer 2: ```suggestion
Learn more about [client diversity](/developers/docs/client-diversity). 
```

What these implementations have in common is they all follow a single specification. Specifications dictate how the Ethereum network and blockchain functions. Every technical detail is defined and specifications can be found as: 
 
- Originally, the [Ethereum Yellow Paper](https://ethereum.github.io/yellowpaper/paper.pdf)
- [Execution specs](https://github.com/ethereum/execution-specs/)
- [Consensus specs](https://github.com/ethereum/consensus-specs)
- [EIPs](https://eips.ethereum.org/) implemented in various [network upgrades](/history/)

> 

Reviewer 2: ```suggestion
```

### Tracking nodes in the network {#network-overview}

There are multiple trackers which offer a real-time overview of nodes in the Ethereum network. Note that due to the nature of decentralized networks, these crawlers can only offer a limited view of the network and might report different results. 

- [Map of nodes](https://etherscan.io/nodetracker) by Etherscan
- [Ethernodes](https://ethernodes.org/) by Bitfly
- [Ethereum Node Crawler](https://crawler.ethereum.org/) 
- [Nodewatch](https://www.nodewatch.io/) by Chainsafe, crawling consensus nodes

## Node types {#node-types}

> If you want to [run your own node](/developers/docs/nodes-and-clients/run-a-node/), you should understand that there are different types of node that consume data differently. In fact, clients can run 3 different types of node - light, full and archive. There are also options of different sync strategies which enable faster synchronization time. Synchronization refers to how quickly it can get the most up-to-date information on Ethereum's state.

Reviewer 2: ```suggestion
If you want to [run your own node](/developers/docs/nodes-and-clients/run-a-node/), you should understand that there are different types of node that consume data differently. In fact, clients can run three different types of node—light, full and archive. There are also options of different sync strategies which enable faster synchronization time. Synchronization refers to how quickly it can get the most up-to-date information on Ethereum's state.
```

Before [The Merge](/upgrades/merge/), consensus and execution layer were separate networks, with all transactions and user activity on the Ethereum happening at what is now the execution layer. One client software provided both execution environment and consensus verification of blocks produced by miners.
The consensus layer, [the Beacon Chain](/upgrades/beacon-chain/), has been running separately since December 2020. It introduced proof-of-stake and coordinated the network of validators based on data from the Ethereum network.

With the Merge, Ethereum transitions to proof-of-stake by connecting these networks. Execution and consensus clients work together to verify Ethereum's state.

Modular design with various software pieces working together is called [encapusalted complexity](https://vitalik.ca/general/2022/02/28/complexity.html). This approach makes it easier to execute The Merge seamlessly and enables the reuse of individual clients, for example, in the [layer 2 ecosystem](/layer-2/).

![Coupled execution and consensus clients](./eth1eth2client.png)
Simplified diagram of a coupled execution and consensus client.

### Client diversity {#client-diversity}

> Both [execution clients](/developers/docs/nodes-and-clients/#consensus-clients) and [consensus clients](/developers/docs/nodes-and-clients/#execution-clients) exist in a variety of programming languages developed by different teams.

Reviewer 4: ```suggestion
Both [execution clients](/developers/docs/nodes-and-clients/#execution-clients) and [consensus clients](/developers/docs/nodes-and-clients/#consensus-clients) exist in a variety of programming languages developed by different teams.
```

Links should be flipped

Learn more about [client diversity](/developers/docs/client-diversity).

What these implementations have in common is they all follow a single specification. Specifications dictate how the Ethereum network and blockchain functions. Every technical detail is defined and specifications can be found as:

- Originally, the [Ethereum Yellow Paper](https://ethereum.github.io/yellowpaper/paper.pdf)
- [Execution specs](https://github.com/ethereum/execution-specs/)
- [Consensus specs](https://github.com/ethereum/consensus-specs)
- [EIPs](https://eips.ethereum.org/) implemented in various [network upgrades](/history/)

### Tracking nodes in the network {#network-overview}

Multiple trackers offer a real-time overview of nodes in the Ethereum network. Note that due to the nature of decentralized networks, these crawlers can only provide a limited view of the network and might report different results.

- [Map of nodes](https://etherscan.io/nodetracker) by Etherscan
- [Ethernodes](https://ethernodes.org/) by Bitfly
- [Ethereum Node Crawler](https://crawler.ethereum.org/)
- [Nodewatch](https://www.nodewatch.io/) by Chainsafe, crawling consensus nodes

## Node types {#node-types}

> If you want to [run your own node](/developers/docs/nodes-and-clients/run-a-node/), you should understand that there are different types of node that consume data differently. In fact, clients can run three different types of node—light, full and archive. There are also options of different sync strategies which enable faster synchronization time. Synchronization refers to how quickly it can get the most up-to-date information on Ethereum's state.

Reviewer 4: ```suggestion
If you want to [run your own node](/developers/docs/nodes-and-clients/run-a-node/), you should understand that there are different types of node that consume data differently. In fact, clients can run three different types of nodes—light, full and archive. There are also options of different sync strategies which enable faster synchronization time. Synchronization refers to how quickly it can get the most up-to-date information on Ethereum's state.
```
Should be plural

### Archive node {#archive-node}

- Stores everything kept in the full node and builds an archive of historical states. It is needed if you want to query something like an account balance at block #4,000,000, or simply and reliably test your own transactions set without mining them using tracing. 
- This data represents units of terabytes, which makes archive nodes less attractive for average users but can be handy for services like block explorers, wallet vendors, and chain analytics.

Syncing clients in any mode other than archive will result in pruned blockchain data. This means, there is no archive of all historical states but the full node is able to build them on demand. 

## Why should I run an Ethereum node? {#why-should-i-run-an-ethereum-node}

Running a node allows you to directly, trustlessly and privately use Ethereum while supporting the network by keeping it more robust and decentralized. 

### Benefits to you {#benefits-to-you}

> Running your own node enables you to use Ethereum in a truly private, self-sufficient and trustless manner. You don't need to trust the network because you can verify the data yourself with your client. "Don't trust, verify" is a popular blockchain mantra.

Author: I believe it is because no information about transactions or addresses is leaked into network. Only IP of the client is 'leaked' to peers but any private data. Later in the text it explains the scenario with a wallet connection to own RPC. It also says 'enables you to use..', it's not supposed to say it automatically does without any effort. Maybe I would specify it better in the initial text

- You can run and self-host other services which depend on data from Ethereum. Most importantly for example Beacon Chain validator but also software like L2 infrastracture, block explorers, payment processors, etc. 
- You can provide your own custom [RPC endpoints](https://ethereum.org/en/developers/docs/apis/json-rpc/). Publicly for the community or even privately hosted Ethereum endpoint enables people to use your node and avoid big centralized providers. 
- You can connect to your node using **Inter-process Communications (IPC)** or rewrite the node to load your program as a plugin. This grants low latency, which helps a lot, e.g. when processing a lot of data using web3 libraries or when you need to replace your transactions as fast as possible (i.e. frontrunning).

> - If you would to stake ETH for securing the proof-of-stake consensus and earning rewards, [solo staking](https://ethereum.org/en/staking/solo/) with your own node is the most impactful and secure approach. 

Author: Agreed! 
I like tags at `Solo home staking` part here https://ethereum.org/en/staking/ and I would like to mention as benefit at least that it's most impactful/trustless/gives a full control

- This data represents units of terabytes, which makes archive nodes less attractive for average users but can be handy for services like block explorers, wallet vendors, and chain analytics.

Syncing clients in any mode other than archive will result in pruned blockchain data. This means, there is no archive of all historical states but the full node is able to build them on demand. 

## Why should I run an Ethereum node? {#why-should-i-run-an-ethereum-node}

Running a node allows you to directly, trustlessly and privately use Ethereum while supporting the network by keeping it more robust and decentralized. 

### Benefits to you {#benefits-to-you}

Running your own node enables you to use Ethereum in a private, self-sufficient and trustless manner. You don't need to trust the network because you can verify the data yourself with your client. "Don't trust, verify" is a popular blockchain mantra.

- Your node verifies all the transactions and blocks against consensus rules by itself. This means you don’t have to rely on any other nodes in the network or fully trust them.

> - You can use an Ethereum wallet with your own node. You can use dapps more securely and privately because you won't have to leak your addresses and balances to random nodes. Everything can be checked with your own client. [MetaMask](https://metamask.io), [Frame](https://frame.sh/) and [many other wallets](/wallets/find-wallet/) can be easily pointed to your own local node. (Check wallets with `RPC importing` filter.)

Reviewer 3: ```suggestion
- You can use an Ethereum wallet with your own node. You can use dapps more securely and privately because you won't have to leak your addresses and balances to random nodes. Everything can be checked with your own client. [MetaMask](https://metamask.io), [Frame](https://frame.sh/), and [many other wallets](/wallets/find-wallet/) offer RPC-importing, allowing them to use your node.
```

> - You can run and self-host other services which depend on data from Ethereum. Most importantly for example Beacon Chain validator but also software like L2 infrastracture, block explorers, payment processors, etc. 

Reviewer 3: ```suggestion
- You can run and self-host other services which depend on data from Ethereum. For example, this might be a Beacon Chain validator, software like layer 2, infrastructure, block explorers, payment processors, etc.  
```

Besu's extensive [documentation](https://besu.hyperledger.org/en/stable/) will guide you trough all details on its features and setups.

### Erigon {#erigon}

Erigon, formerly known as Turbo‐Geth, started as a fork of Go Ethereum oriented toward speed and disk‐space efficiency. Erigon is a completely re-architected implementation of Ethereum, currently written in Go but with implementations in other languages in work, e.g. [Akula](https://medium.com/@vorot93/meet-akula-the-fastest-ethereum-implementation-ever-built-58eaca244c39). Erigon's goal is to provide a faster, more modular, and more optimized implementation of Ethereum. It can perform a full archive node sync using around 2TB of disk space, in under 3 days.

### Go Ethereum {#geth}

Go Ethereum (Geth for short) is one of the original implementations of the Ethereum protocol. Currently, it is the most widespread client with the biggest user base and variety of tooling for users and developers. It is written in Go, fully open source and licensed under the GNU LGPL v3.

Learn more about Geth in its [documentation](https://geth.ethereum.org/docs/).

### Nethermind {#nethermind}

> Nethermind is an Ethereum implementation created with the C# .NET tech stack, licensed with LGPL-3.0, running on all major platforms including ARM. It offers great performance with:

Reviewer 4: ```suggestion
Nethermind is an Ethereum implementation created with the C# .NET tech stack, licensed with GNU LGPL-3.0, running on all major platforms including ARM. It offers great performance with:
```

Should this also have GNU?

Reviewer 4: I think this is the same license as geth above, should be consistent.

| Client                                                   | Language | Operating systems     | Networks                                   | Sync strategies              | State pruning   |
| -------------------------------------------------------- | -------- | --------------------- | ------------------------------------------ | ---------------------------- | --------------- |
| [Akula](https://akula.app)                               | Rust     | Linux                 | Mainnet, Görli, Rinkeby, Ropsten, and more | Full                         | Archive, Pruned |
| [Besu](https://pegasys.tech/solutions/hyperledger-besu/) | Java     | Linux, Windows, macOS | Mainnet, Rinkeby, Ropsten, Görli, and more | Fast, Full, Snap, Checkpoint | Archive, Pruned |
| [Erigon](https://github.com/ledgerwatch/erigon)          | Go       | Linux, Windows, macOS | Mainnet, Görli, Rinkeby, Ropsten           | Full, Snap                   | Archive, Pruned |
| [Geth](https://geth.ethereum.org/)                       | Go       | Linux, Windows, macOS | Mainnet, Görli, Rinkeby, Ropsten           | Snap, Full                   | Archive, Pruned |

> | [Nethermind](http://nethermind.io/)                      | C#, .NET | Linux, Windows, macOS | Mainnet, Görli, Ropsten, Rinkeby, and more | Snap, Fast, Beam             | Archive, Pruned |

Reviewer 3: Does Nethermind have beam sync still?

Author: Will remove

- an optimized virtual machine
- state access
- networking and rich features like Prometheus/Grafana dashboards, seq enterprise logging support, JSON RPC tracing, and analytics plugins.

Nethermind also has [detailed documentation](https://docs.nethermind.io), strong dev support, an online community and 24/7 support available for premium users.

## Consensus clients (formerly 'Eth2' clients) {#consensus-clients}

There are multiple consensus clients (previously known as 'Eth2' clients) to support the [consensus upgrades](/upgrades/beacon-chain/). They are running the Beacon Chain and will provide a proof-of-stake consensus mechanism to execution clients after [The Merge](/upgrades/merge/).

[View consensus clients](/upgrades/get-involved/#clients).

| Client                                                        | Language   | Operating systems     | Networks                              |
| ------------------------------------------------------------- | ---------- | --------------------- | ------------------------------------- |

> | [Lighthouse](https://lighthouse.sigmaprime.io/)               | Rust       | Linux, Windows, macOS | Beacon Chain, Prater, Pyrmont         |
> | [Lodestar](https://lodestar.chainsafe.io/)                    | TypeScript | Linux, Windows, macOS | Beacon Chain, Prater                  |
> | [Nimbus](https://nimbus.team/)                                | Nim        | Linux, Windows, macOS | Beacon Chain, Prater                  |
> | [Prysm](https://docs.prylabs.network/docs/getting-started/)   | Go         | Linux, Windows, macOS | Beacon Chain, Gnosis, Prater, Pyrmont |
> | [Teku](https://consensys.net/knowledge-base/ethereum-2/teku/) | Java       | Linux, Windows, macOS | Beacon Chain, Gnosis, Prater, Sepolia |

Reviewer 2: We should probably update mentions of "Prater" to "Goerli" now (🎉 )
```suggestion
| [Lighthouse](https://lighthouse.sigmaprime.io/)               | Rust       | Linux, Windows, macOS | Beacon Chain, Goerli, Pyrmont         |
| [Lodestar](https://lodestar.chainsafe.io/)                    | TypeScript | Linux, Windows, macOS | Beacon Chain, Goerli                  |
| [Nimbus](https://nimbus.team/)                                | Nim        | Linux, Windows, macOS | Beacon Chain, Goerli                  |
| [Prysm](https://docs.prylabs.network/docs/getting-started/)   | Go         | Linux, Windows, macOS | Beacon Chain, Gnosis, Goerli, Pyrmont |
| [Teku](https://consensys.net/knowledge-base/ethereum-2/teku/) | Java       | Linux, Windows, macOS | Beacon Chain, Gnosis, Goerli, Sepolia |
```

### Lighthouse

Lighthouse is a consensus client implementation written in Rust under the Apache-2.0 license. It is maintained by Sigma Prime and has been stable and production-ready since Beacon Chain genesis. It is relied upon by a wide variety of enterprises, staking pools and individuals. It aims to be secure, performant and interoperable in a wide range of environments from desktop PCs through to sophisticated automated deployments.

Documentation can be found in [Lighthouse Book](https://lighthouse-book.sigmaprime.io/).

### Lodestar

> Lodestar is a production-ready consensus client implementation written in Typescript under the LGPL-3.0 license. It is maintained by ChainSafe Systems and is the newest of the consensus clients for solo-stakers, developers and researchers. Lodestar consists of a beacon node and validator client powered by JavaScript implementations of Ethereum protocols. We aim to improve Ethereum usability with light clients, expand accessibility to a larger group of developers and further contribute to ecosystem diversity.

Reviewer 2: ```suggestion
Lodestar is a production-ready consensus client implementation written in Typescript under the LGPL-3.0 license. It is maintained by ChainSafe Systems and is the newest of the consensus clients for solo-stakers, developers and researchers. Lodestar consists of a beacon node and validator client powered by JavaScript implementations of Ethereum protocols. Lodestar aims to improve Ethereum usability with light clients, expand accessibility to a larger group of developers and further contribute to ecosystem diversity.
```

More information can be found on our [Lodestar website](https://lodestar.chainsafe.io/)

### Nimbus

Nimbus is a consensus client implementation written in Nim under Apache-2.0 license. It is a production-ready client in use by solo-stakers and staking pools. It is designed for resource efficiency, making it easy to run Nimbus on resource-restricted devices and enterprise infrastructure with equal ease, without compromising stability or reward performance. A lighter resource footprint means the client has greater margin-of-safety when the network is under stress. 

Documentation can be found in the [Nimbus Guide](https://nimbus.guide/).

### Prysm

Prysm is a full-featured, open source consensus client written in Go under the GPL-3.0 license. It features an optional webapp UI and prioritizes user experience, documentation, and configurability for both stake-at-home and institutional users.

Visit [Prysm docs](https://docs.prylabs.network/docs/getting-started/) to learn more. 

### Teku

> Teku is one of the original beacon chain genesis clients. Alongside the usual goals (security, robustness, stability, usability, performance) it specifically aims to comply fully with all the various consensus client standards.

Reviewer 2: ```suggestion
Teku is one of the original Beacon Chain genesis clients. Alongside the usual goals (security, robustness, stability, usability, performance) it specifically aims to comply fully with all the various consensus client standards.
```

- With an increasing number of transactions, it can take days to weeks to process all transactions.

#### Fast sync

Fast sync downloads all blocks (including headers, transactions, and receipts), verifies all headers, downloads the state and verifies it against the headers.

- Relies on the security of the consensus mechanism.
- Synchronization takes only a few hours.

#### Light sync

Light client mode downloads all block headers, block data, and verifies some randomly. Only syncs tip of the chain from the trusted checkpoint.

- Gets only the latest state while relying on trust in developers and consensus mechanism.
- Client ready to use with current network state in a few minutes.

[More on Light clients](https://www.parity.io/blog/what-is-a-light-client/)

#### Snap sync

The latest approach to syncing the client, pioneered by Geth. Using dynamic snapshots served by peers retrieves all the account and storage data without downloading intermediate trie nodes and then reconstructs the Merkle trie locally.

- Fastest sync strategy, currently default in Ethereum mainnet 
- Saves a lot of disk usage and network bandwidth without sacrificing security

> [More on Snap sync](https://github.com/ethereum/devp2p/blob/master/caps/snap.md)

Reviewer 2: ```suggestion
[More on snap sync](https://github.com/ethereum/devp2p/blob/master/caps/snap.md)
```

#### Consensus layer sync modes

#### Optimistic sync

Optimistic sync is a post-merge synchronization strategy designed to be opt-in and backwards compatible, allowing execution nodes to sync via established methods. The execution engine can *optimistically* import beacon blocks without fully verifying them, find the latest head, and then start syncing the chain with the above methods. Then, after the execution client has caught up, it will inform the consensus client of the validity of the transactions in the Beacon Chain.

> [More on Optimistic sync](https://github.com/ethereum/consensus-specs/blob/dev/sync/optimistic.md)

Reviewer 2: ```suggestion
[More on optimistic sync](https://github.com/ethereum/consensus-specs/blob/dev/sync/optimistic.md)
```

#### Checkpoint sync

Also called weak subjectivity sync, this strategy creates a superior user experience for syncing Beacon Node. It's based on assumptions of [weak subjectivity](/developers/docs/consensus-mechanisms/pos/weak-subjectivity/) which enables syncing Beacon Chain from a recent weak subjectivity checkpoint instead of genesis. This makes the initial sync time significantly faster with similiar trust assumptions as syncing from genesis. 

In practice, this means your node connects to a remote service to download recent finalized states and continues verifying data from that point. Third party providing the data is trusted and should be picked carefully. 

> More on [Checkpoint sync](https://notes.ethereum.org/@djrtwo/ws-sync-in-practice)

Reviewer 2: ```suggestion
More on [checkpoint sync](https://notes.ethereum.org/@djrtwo/ws-sync-in-practice)
```

Besu's extensive [documentation](https://besu.hyperledger.org/en/stable/) will guide you trough all details on its features and setups. 

#### Erigon {#erigon}

Erigon, formerly known as Turbo‐Geth, started as a fork of Go Ethereum oriented toward speed and disk‐space efficiency. Erigon is a completely re-architected implementation of Ethereum, currently written in Go but with implementations in other languages in work, e.g. [Akula](https://medium.com/@vorot93/meet-akula-the-fastest-ethereum-implementation-ever-built-58eaca244c39). Erigon's goal is to provide a faster, more modular, and more optimized implementation of Ethereum. It can perform a full archive node sync using around 2TB of disk space, in under 3 days. 

## Consensus clients (formerly 'Eth2' clients) {#consensus-clients}

There are multiple consensus clients (previously known as 'Eth2' clients) to support the [consensus upgrades](/upgrades/beacon-chain/). They are running the Beacon Chain and will provide a proof-of-stake consensus mechanism to execution clients after [The Merge](/upgrades/merge/).

[View consensus clients](/upgrades/get-involved/#clients).

> | Client                                                        | Language   | Operating systems     | Networks                              |

Reviewer 5: I would order them alphabetically, and we can mention that in the table header

Teku offers very flexible deployment options. The beacon node and validator client can be run together as a single process, which is extremely convenient for solo stakers, or they can be run separately for sophisticated staking operations. In addition, Teku is fully interoperable with [Web3Signer](https://github.com/ConsenSys/web3signer/) for signing key security and slashing protection.

Teku is written in Java and is Apache 2.0 licensed. It is developed by the Protocols team at ConsenSys that is also responsible for Besu and Web3Signer.

### Nimbus

Nimbus is a consensus client implementation written in Nim under Apache-2.0 license. It is a production-ready client in use by solo-stakers and staking pools. It is designed for resource efficiency, making it easy to run Nimbus on resource-restricted devices and enterprise infrastructure with equal ease, without compromising stability or reward performance. A lighter resource footprint means the client has greater margin-of-safety when the network is under stress. 

> Documentation can be found in the Nimbus Guide: https://nimbus.guide/

Reviewer 5: Can directly link vs. exposing the URL

Visit [Prysm docs](https://docs.prylabs.network/docs/getting-started/) to learn more. 

### Synchronization modes {#sync-modes}

To follow and verify current data in the network, the Ethereum client needs to sync with the latest network state. This is done by downloading data from peers, cryptographically verifying their integrity, and building a local blockchain database.

Synchronization modes represent different approaches to this process with various trade-offs. Clients also vary in their implementation of sync algorithms. Always refer to the official documentation of your chosen client for specifics on implementation.

#### Full sync {#full-sync}

Full sync downloads all blocks (including headers, transactions, and receipts) and generates the state of the blockchain incrementally by executing every block from genesis.

- Minimizes trust and offers the highest security by verifying every transaction.
- With an increasing number of transactions, it can take days to weeks to process all transactions.

> #### Fast sync

Reviewer 5: I would mention that Snap Sync is more commonly used now, and perhaps switch their order in the page.

Fast sync downloads all blocks (including headers, transactions, and receipts), verifies all headers, downloads the state and verifies it against the headers.

- Relies on the security of the consensus mechanism.
- Synchronization takes only a few hours.

#### Light sync

Light client mode downloads all block headers, block data, and verifies some randomly. Only syncs tip of the chain from the trusted checkpoint.

- Gets only the latest state while relying on trust in developers and consensus mechanism.
- Client ready to use with current network state in a few minutes.

[More on Light clients](https://www.parity.io/blog/what-is-a-light-client/)

#### Snap sync

The latest approach to syncing the client, pioneered by Geth. Using dynamic snapshots served by peers retrieves all the account and storage data without downloading intermediate trie nodes and then reconstructs the Merkle trie locally.

- Fastest sync strategy, currently default in Geth and Nethermind 
- Saves a lot of disk usage and network bandwidth without sacrificing security.

> [More on Snap](https://github.com/ethereum/devp2p/blob/master/caps/snap.md)

Reviewer 5: ```suggestion
[More on Snap sync](https://github.com/ethereum/devp2p/blob/master/caps/snap.md)
```

> #### Beam sync

Reviewer 5: Is this still supported in Nethermind? If not, I'd remove it as it was never production ready IIRC

Implemented by Nethermind and Trinity. Works like fast sync but also downloads the data needed to execute latest blocks, which allows you to query the chain within the first few minutes from starting.

- Syncs state first and enables you to query RPC in a few minutes.
- Still in development and not fully reliable, background sync is slowed down and RPC responses might fail.

[More on Beam](https://medium.com/@jason.carver/intro-to-beam-sync-a0fd168be14a)

> #### Optimistic sync

Reviewer 5: This is a bit odd to mention here given it's the only CL sync strategy. Perhaps if we want to also mention CL sync we should more clearly break down the page into CL/EL sync modes and also include other CL sync modes?

Author: Agreed, I can break it down to two parts and should also add Checkpoint sync. Any other for CL?
